typing/typecore.ml

```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+        and optional = is_optional l in
+        let use_arg sarg l' =
+          let wrapped_in_some = optional && not (is_optional l') in
+          if wrapped_in_some then
+            may_warn sarg.pexp_loc
+              (not_principal "using an optional argument here");
+          Arg (Known_arg { sarg; ty_arg; ty_arg0; wrapped_in_some })
+        in
+        let eliminate_optional_arg () =
+          may_warn funct.exp_loc
+            (Warnings.Non_principal_labels "eliminated optional argument");
+          Arg (Eliminated_optional_arg { ty_arg; level = lv })
+        in
+        let remaining_sargs, arg =
+          if ignore_labels then begin
+            (* No reordering is allowed, process arguments in order *)
+            match sargs with
+            | [] -> assert false
+            | (l', sarg) :: remaining_sargs ->
+                if name = label_name l' || (not optional && l' = Nolabel) then
+                  (remaining_sargs, use_arg sarg l')
+                else if
+                  optional &&
+                  not (List.exists (fun (l, _) -> name = label_name l)
+                         remaining_sargs) &&
+                  List.exists (function (Nolabel, _) -> true | _ -> false)
+                    sargs
+                then
+                  (sargs, eliminate_optional_arg ())
+                else
+                  raise(Error(sarg.pexp_loc, env,
+                              Apply_wrong_label(l', ty_fun', optional)))
+          end else
+            (* Arguments can be commuted, try to fetch the argument
+               corresponding to the first parameter. *)
+            match extract_label name sargs with
+            | Some (l', sarg, commuted, remaining_sargs) ->
+                if commuted then begin
+                  may_warn sarg.pexp_loc
+                    (not_principal "commuting this argument")
+                end;
+                if not optional && is_optional l' then
+                  Location.prerr_warning sarg.pexp_loc
+                    (Warnings.Nonoptional_label (Asttypes.string_of_label l));
+                remaining_sargs, use_arg sarg l'
+            | None ->
+                sargs,
+                if optional && List.mem_assoc Nolabel sargs then
+                  eliminate_optional_arg ()
+                else begin
+                  (* No argument was given for this parameter, we abstract over
+                     it. *)
+                  may_warn funct.exp_loc
+                    (Warnings.Non_principal_labels "commuted an argument");
+                  Omitted { ty_arg; level = lv }
+                end
+        in
+        loop ty_ret ty_ret0 ((l, arg) :: rev_args) remaining_sargs
+    | _ ->
+      (* We're not looking at a *known* function type anymore, or there are no
+        arguments left. *)
+      collect_unknown_apply_args env funct ty_fun0 rev_args sargs
+  in
+  loop ty_fun ty_fun0 [] sargs
+
+let type_omitted_parameters ty_ret args =
```

Comment:
Still trying to map the new code to the old code: it is unclear to me where this function comes from, it appears to be new. There seems to be some redudancy with the function `remaining_function_type` above, I am not sure why this is.

Reply:
Part of the code for this one actually comes from the `result_type` auxiliary function defined at the beginning of trunk's `type_application` function and which is called other the list of `!omitted_parameters`.

Reply:
But `result_type` was already moved into `remaining_function_type` above. Why have the same Tarrow-building logic in two different places?

Reply:
For me the difference is that :
- `remaining_function_type` of `(f : ?x:int -> y:bool -> float -> t) 3.` is `?x:int -> y:bool -> t` because `?x` will go in the `Arg (Eliminated_optional_arg _)` case.
- `type_omitted_parameters` of `(f : ?x:int -> y:bool -> float -> t) 3.` is `y:bool -> t`
The first case is used for error messages when the second is the resulting type after eta-expanding ommited labelled arguments.
It might be nice to put both function together with a comment to explain why they are different.

Reply:
We could merge the two function by adding an `?reinject_eliminated_optional_arg` arguments to `type_omitted_parameters` and discarding the argument part in the error path that was using `remaining_function_type`.

I also think that it could be nice the rename the function to `return_function_type` or `return_type_with_omitted_parameters` to better convey the idea that there is no typing going on in the function `type_omitted_parameters`.

Reply:
The reason it is called `type_omitted_parameters` is that it maps over the omitted parameters turning them from their untyped form -- `untyped_omitted_param` -- into their typed form. Unfortunately, in this patch the typed form is just `unit` so it looks a bit weird. Our work on both modes and kinds requires us to put more information into the typed form, so that we have a type `omitted_param` in `Typedtree` and that is what is produced for the omitted parameters here.

I do think this is still to right way to think about this function however, so I'd prefer if this view could be preserved in the name.

It is true than this function is really a `fold_map` rather than a `map` -- it also computes the `result_type` as it goes -- and this isn't reflected in the current name. It probably would be better if the name included that detail. How about something like `type_omitted_parameters_and_build_result_type`?

Reply:
Here I really think that this PR should use a name that corresponds to what is done, not what will be done in the future. You can easily change this name when the meaning changes: this function is used only once.

Reply:
My understanding is that it's not just about a change in the future, it about reducing the diff between upstream and Jane Street's _existing_ fork of the type-checker codebase. I think that it's worth accommodating this need if we can at reasonable cost, to make such upstreaming easier (and worthy) in the future. Maybe we could find a name that would be acceptable in both versions, for example `handle_omittted_parameters_and_build_result_type`?

Reply:
My point was actually slightly stronger than just about keeping the diff small. I was also arguing that semantically this is the place where we "type" the omitted arguments: taking them from their representation in the parsetree to their representation in the typedtree. Just because the representation in the typedtree is `()` doesn't mean it is the wrong way to picture this function.

Still, I would settle for `handle_omitted_parameters_...` if people object to using the word `type_` for this.